core/src/main/java/kafka/server/ServerSocketFactory.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka.server;
+
+import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.common.network.Selectable;
+import org.apache.kafka.common.utils.Utils;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.SocketException;
+import java.nio.channels.ServerSocketChannel;
+
+public interface ServerSocketFactory {
+    ServerSocketChannel openServerSocket(
+        String listenerName,
+        InetSocketAddress socketAddress,
+        int listenBacklogSize,
+        int recvBufferSize
+    ) throws IOException;
+
+    class KafkaServerSocketFactory implements ServerSocketFactory {
+        public static final KafkaServerSocketFactory INSTANCE = new KafkaServerSocketFactory();
+
+        @Override
+        public ServerSocketChannel openServerSocket(
+                String listenerName,
+                InetSocketAddress socketAddress,
+                int listenBacklogSize,
+                int recvBufferSize
+        ) throws IOException {
+            ServerSocketChannel socketChannel = ServerSocketChannel.open();
+            try {
+                socketChannel.configureBlocking(false);
+                if (recvBufferSize != Selectable.USE_DEFAULT_BUFFER_SIZE) {
+                    socketChannel.socket().setReceiveBufferSize(recvBufferSize);
+                }
+                socketChannel.socket().bind(socketAddress, listenBacklogSize);
+            } catch (SocketException e) {
+                Utils.closeQuietly(socketChannel, "server socket");
+                throw new KafkaException(String.format("Socket server failed to bind to %s:%d: %s.",
```

Comment:
I did it this way to keep the same behavior as the existing `openServerSocket` function, which doesn't catch and wrap `IOException` when it's raised from `ServerSocketChannel.open()`, but does catch and wrap it when it's reaised from `socket.bind()` and subsequent calls.

If we want to change the behavior, it would be better to have a follow-on JIRA I think...